### ISSUE TITLE
Size/density calculation with mm

### ISSUE THREAD

**Issue by klabarge** [qzind/tray#189]
If the `density` is not specified when using `units: 'mm'`, HTML printing does not work on macOS. The `size` or `density` is incorrectly calculated. This is only reproducible on macOS

The log indicates that these dimensions are set:
* Width: 172719 MM (6799 IN)
* Height: 52069 MM (2049 IN)

### Environment
**macOS:** 10.12.4 (16E195)
**QZ Tray:** Reprocuded on 2.02 and 2.03

### Steps

1.  Go to [demo page](https://demo.qz.io) > Pixel Printing tab
1.  Change the Units radio button to "MM"
1. Try to print using the "Print HTML" button
    * Observe the print job fail (stack trace below)
1.  Change the density to 23.622 (this is 600dpi)
    * HTML printing now succeeds

```
[DEBUG] 2017-04-12 21:15:16,987 @ qz.ws.PrintSocketClient:?
	Message: {"call":"print","promise":{},"params":{"printer":{"name":"PDFwriter"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":600,"interpolation":"","jobName":"","margins":"0","orientation":"","paperThickness":"","printerTray":"","rasterize":true,"rotation":"0","scaleContent":true,"size":null,"units":"mm","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":[{"type":"html","format":"plain","data":"<html>   <table style=\"font-family: monospace; border: 1px;\">       <tr style=\"height: 6cm;\">           <td valign=\"top\"><h2>*&nbsp; QZ Print Plugin HTML Printing &nbsp;*</h2><span style=\"color: #F00;\">Version:</span> 2.0.3<br/><span style=\"color: #F00;\">Visit:</span> https://qz.io/</td>           <td valign=\"top\"><img src=\"https://demo.qz.io/assets/img/image_sample.png\"></td>       </tr>   </table></html>"}]},"timestamp":1492046116986,"uid":"p6lc9h"}
[WARN] 2017-04-12 21:15:16,989 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2017-04-12 21:15:17,919 @ qz.common.TrayManager:?
	Allowed localhost to print to PDFwriter
[DEBUG] 2017-04-12 21:15:18,060 @ qz.utils.PrintingUtilities:?
	Allowing 3 simultaneous processors based on memory available (1908 MB)
[TRACE] 2017-04-12 21:15:18,060 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/3 already in use
[DEBUG] 2017-04-12 21:15:18,066 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintHTML to print
[DEBUG] 2017-04-12 21:15:18,067 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching PDFwriter
[DEBUG] 2017-04-12 21:15:18,067 @ qz.printer.PrintServiceMatcher:?
	Found 3 printers
[DEBUG] 2017-04-12 21:15:18,074 @ qz.printer.PrintServiceMatcher:?
	Found match: PDFwriter
[WARN] 2017-04-12 21:15:18,078 @ qz.printer.PrintOptions:?
	Cannot read  as a double for density, using default
[WARN] 2017-04-12 21:15:18,078 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for interpolation, using default
[WARN] 2017-04-12 21:15:18,079 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for orientation, using default
[WARN] 2017-04-12 21:15:18,080 @ qz.printer.PrintOptions:?
	Cannot read  as a double for paperThickness, using default
[DEBUG] 2017-04-12 21:15:18,084 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -a]
[DEBUG] 2017-04-12 21:15:18,103 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -l, -p, zebra]
[INFO] 2017-04-12 21:15:18,121 @ qz.utils.ShellUtilities:?
	zebra: zebra
[DEBUG] 2017-04-12 21:15:18,121 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -l, -p, zebra_class]
[INFO] 2017-04-12 21:15:18,139 @ qz.utils.ShellUtilities:?
	zebra_class: zebra
[DEBUG] 2017-04-12 21:15:18,139 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -l, -p, PDFwriter]
[INFO] 2017-04-12 21:15:18,159 @ qz.utils.ShellUtilities:?
	PDFwriter: PDFwriter
[DEBUG] 2017-04-12 21:15:18,159 @ qz.utils.ShellUtilities:?
	Executing: [lpstat, -l, -p, Zebra_epl]
[INFO] 2017-04-12 21:15:18,178 @ qz.utils.ShellUtilities:?
	Zebra_epl: Zebra_epl
[DEBUG] 2017-04-12 21:15:18,178 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, zebra, -l]
[DEBUG] 2017-04-12 21:15:18,216 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, zebra_class, -l]
[DEBUG] 2017-04-12 21:15:18,236 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PDFwriter, -l]
[DEBUG] 2017-04-12 21:15:18,257 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Zebra_epl, -l]
[TRACE] 2017-04-12 21:15:18,367 @ qz.printer.action.WebApp:?
	Waiting for JavaFX..
[DEBUG] 2017-04-12 21:15:18,499 @ qz.printer.action.WebApp:?
	Started JavaFX
[TRACE] 2017-04-12 21:15:19,369 @ qz.printer.action.WebApp:?
	Waiting for capture..
[TRACE] 2017-04-12 21:15:19,509 @ qz.printer.action.WebApp:?
	Done: -1.0 > 0.0
[TRACE] 2017-04-12 21:15:19,509 @ qz.printer.action.WebApp:?
	New state: READY > SCHEDULED
[TRACE] 2017-04-12 21:15:19,509 @ qz.printer.action.WebApp:?
	New state: SCHEDULED > RUNNING
[TRACE] 2017-04-12 21:15:20,079 @ qz.printer.action.WebApp:?
	Done: 0.0 > 35.24521681009274
[TRACE] 2017-04-12 21:15:20,103 @ qz.printer.action.WebApp:?
	Done: 35.24521681009274 > 100.0
[TRACE] 2017-04-12 21:15:20,103 @ qz.printer.action.WebApp:?
	New state: RUNNING > SUCCEEDED
[TRACE] 2017-04-12 21:15:20,136 @ qz.printer.action.WebApp:?
	Zooming in by x211.6666634877523 for increased quality
[TRACE] 2017-04-12 21:15:20,136 @ qz.printer.action.WebApp:?
	Setting HTML page width to 172719.99740600586
java.lang.RuntimeException: Requested texture dimensions (172720x100) require dimensions (0x100) that exceed maximum texture size (16384)
	at com.sun.prism.es2.ES2RTTexture.create(ES2RTTexture.java:220)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:158)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:154)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.createTexture(WCPageBackBufferImpl.java:52)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.validate(WCPageBackBufferImpl.java:114)
	at com.sun.webkit.WebPage.paint(WebPage.java:656)
	at com.sun.javafx.sg.prism.web.NGWebView.renderContent(NGWebView.java:96)
	at com.sun.javafx.sg.prism.NGNode.doRender(NGNode.java:2053)
	at com.sun.javafx.sg.prism.NGNode.render(NGNode.java:1945)
	at com.sun.javafx.tk.quantum.ViewPainter.doPaint(ViewPainter.java:477)
	at com.sun.javafx.tk.quantum.ViewPainter.paintImpl(ViewPainter.java:330)
	at com.sun.javafx.tk.quantum.PresentingPainter.run(PresentingPainter.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at com.sun.javafx.tk.RenderJob.run(RenderJob.java:58)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:125)
	at java.lang.Thread.run(Thread.java:745)
java.lang.RuntimeException: Requested texture dimensions (172720x100) require dimensions (0x100) that exceed maximum texture size (16384)
	at com.sun.prism.es2.ES2RTTexture.create(ES2RTTexture.java:220)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:158)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:154)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.createTexture(WCPageBackBufferImpl.java:52)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.validate(WCPageBackBufferImpl.java:114)
	at com.sun.webkit.WebPage.paint(WebPage.java:656)
	at com.sun.javafx.sg.prism.web.NGWebView.renderContent(NGWebView.java:96)
	at com.sun.javafx.sg.prism.NGNode.doRender(NGNode.java:2053)
	at com.sun.javafx.sg.prism.NGNode.render(NGNode.java:1945)
	at com.sun.javafx.tk.quantum.ViewPainter.doPaint(ViewPainter.java:477)
	at com.sun.javafx.tk.quantum.ViewPainter.paintImpl(ViewPainter.java:330)
	at com.sun.javafx.tk.quantum.PresentingPainter.run(PresentingPainter.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at com.sun.javafx.tk.RenderJob.run(RenderJob.java:58)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:125)
	at java.lang.Thread.run(Thread.java:745)
[TRACE] 2017-04-12 21:15:20,241 @ qz.printer.action.WebApp:?
	Setting HTML page height to 52069.99921798706
java.lang.RuntimeException: Requested texture dimensions (172720x52070) require dimensions (0x0) that exceed maximum texture size (16384)
	at com.sun.prism.es2.ES2RTTexture.create(ES2RTTexture.java:220)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:158)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:154)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.createTexture(WCPageBackBufferImpl.java:52)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.validate(WCPageBackBufferImpl.java:114)
	at com.sun.webkit.WebPage.paint(WebPage.java:656)
	at com.sun.javafx.sg.prism.web.NGWebView.renderContent(NGWebView.java:96)
	at com.sun.javafx.sg.prism.NGNode.doRender(NGNode.java:2053)
	at com.sun.javafx.sg.prism.NGNode.render(NGNode.java:1945)
	at com.sun.javafx.tk.quantum.ViewPainter.doPaint(ViewPainter.java:477)
	at com.sun.javafx.tk.quantum.ViewPainter.paintImpl(ViewPainter.java:330)
	at com.sun.javafx.tk.quantum.PresentingPainter.run(PresentingPainter.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at com.sun.javafx.tk.RenderJob.run(RenderJob.java:58)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:125)
	at java.lang.Thread.run(Thread.java:745)
java.lang.RuntimeException: Requested texture dimensions (172720x52070) require dimensions (0x0) that exceed maximum texture size (16384)
	at com.sun.prism.es2.ES2RTTexture.create(ES2RTTexture.java:220)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:158)
	at com.sun.prism.es2.ES2ResourceFactory.createRTTexture(ES2ResourceFactory.java:154)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.createTexture(WCPageBackBufferImpl.java:52)
	at com.sun.javafx.webkit.prism.WCPageBackBufferImpl.validate(WCPageBackBufferImpl.java:114)
	at com.sun.webkit.WebPage.paint(WebPage.java:656)
	at com.sun.javafx.sg.prism.web.NGWebView.renderContent(NGWebView.java:96)
	at com.sun.javafx.sg.prism.NGNode.doRender(NGNode.java:2053)
	at com.sun.javafx.sg.prism.NGNode.render(NGNode.java:1945)
	at com.sun.javafx.tk.quantum.ViewPainter.doPaint(ViewPainter.java:477)
	at com.sun.javafx.tk.quantum.ViewPainter.paintImpl(ViewPainter.java:330)
	at com.sun.javafx.tk.quantum.PresentingPainter.run(PresentingPainter.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at com.sun.javafx.tk.RenderJob.run(RenderJob.java:58)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at com.sun.javafx.tk.quantum.QuantumRenderer$PipelineRunnable.run(QuantumRenderer.java:125)
	at java.lang.Thread.run(Thread.java:745)
[DEBUG] 2017-04-12 21:15:20,357 @ qz.printer.action.WebApp:?
	Attempting image capture
[TRACE] 2017-04-12 21:15:20,585 @ qz.printer.action.WebApp:?
	Waiting for capture..
[ERROR] 2017-04-12 21:15:21,607 @ qz.utils.PrintingUtilities:?
	Failed to print
java.lang.UnsupportedOperationException: Cannot parse (PLAIN)<html>   <table style="font-family: monospace; border: 1px;">       <tr style="height: 6cm;">           <td valign="top"><h2>*&nbsp; QZ Print Plugin HTML Printing &nbsp;*</h2><span style="color: #F00;">Version:</span> 2.0.3<br/><span style="color: #F00;">Visit:</span> https://qz.io/</td>           <td valign="top"><img src="https://demo.qz.io/assets/img/image_sample.png"></td>       </tr>   </table></html> as HTML
	at qz.printer.action.PrintHTML.parseData(Unknown Source)
	at qz.utils.PrintingUtilities.processPrintRequest(Unknown Source)
	at qz.ws.PrintSocketClient.processMessage(Unknown Source)
	at qz.ws.PrintSocketClient.onMessage(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
	at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.lang.OutOfMemoryError: Java heap space
	at qz.printer.action.WebApp.capture(Unknown Source)
	... 14 more
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:57)
	at java.nio.ByteBuffer.allocate(ByteBuffer.java:335)
	at com.sun.javafx.tk.quantum.QuantumToolkit.createPlatformImage(QuantumToolkit.java:1343)
	at javafx.scene.image.Image.<init>(Image.java:732)
	at javafx.scene.image.WritableImage.<init>(WritableImage.java:75)
	at javafx.scene.Scene.doSnapshot(Scene.java:1194)
	at javafx.scene.Node.doSnapshot(Node.java:1862)
	at javafx.scene.Node.snapshot(Node.java:1940)
	at qz.printer.action.WebApp$4$1.handle(Unknown Source)
	at qz.printer.action.WebApp$4$1.handle(Unknown Source)
	at javafx.animation.Animation.impl_finished(Animation.java:1132)
	at javafx.animation.AnimationAccessorImpl.finished(AnimationAccessorImpl.java:49)
	at com.sun.scenario.animation.shared.SingleLoopClipEnvelope.timePulse(SingleLoopClipEnvelope.java:103)
	at javafx.animation.Animation.impl_timePulse(Animation.java:1102)
	at javafx.animation.Animation$1.lambda$timePulse$26(Animation.java:186)
	at javafx.animation.Animation$1$$Lambda$208/1141168968.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at javafx.animation.Animation$1.timePulse(Animation.java:185)
	at com.sun.scenario.animation.AbstractMasterTimer.timePulseImpl(AbstractMasterTimer.java:344)
	at com.sun.scenario.animation.AbstractMasterTimer$MainLoop.run(AbstractMasterTimer.java:267)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:506)
	at com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:490)
	at com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$405(QuantumToolkit.java:319)
	at com.sun.javafx.tk.quantum.QuantumToolkit$$Lambda$78/293235145.run(Unknown Source)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
[TRACE] 2017-04-12 21:15:21,611 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

**Comment by tresf** [comment]
@bberenz I think we have a bug with fallback density.  E.g. setting 600 never converts from `dpi` to `dpmm` making the zoom explode.

https://github.com/qzind/tray/blob/2.0/src/qz/printer/PrintOptions.java#L211-L221

**Comment by bberenz** [comment]
Should be fixed under f488bd3a37ac07f30ea3240cfcc4fbfdf954bac0

We had explicitly defaulted the javascript density fallback to 600 (assuming dpi), but this value was meant to be in relation to the selected units, so when dpmm was chosen and we had to use the fallback, we assumed the user purposely meant 600dpmm. I've set it to null by default so that users will have to set it in the config themselves (in their selected units) to use it. Otherwise we use the java fallback (so the fallback in case the fallback fails), which is 600dpi (regardless of selected units) and will be correctly converted to the expected unit's density before passing to printing.

So basically, everything still works exactly as before - we just end up using a different fallback point for the defaults, so that the user configuration values will work as expected with the selected units.

**Comment by tresf** [comment]
Thanks @bberenz.

@klabarge can you unit test this against stable and confirm inches and mm are the same.  Please look at the logs for confirmation.

**Comment by klabarge** [comment]
> Otherwise we use the java fallback (so the fallback in case the fallback fails), which is 600dpi (regardless of selected units) and will be correctly converted to the expected unit's density before passing to printing.

This does work, however the DPI that is set for `mm` is 609.6, resulting in a slightly larger image. Is this acceptable?



#### Units in MM
```
[TRACE] 2017-04-25 19:38:59,787 @ qz.printer.action.WebApp:?
	Zooming in by x8.466666539510092 for increased quality
[TRACE] 2017-04-25 19:38:59,787 @ qz.printer.action.WebApp:?
	Setting HTML page width to 6908.799896240235
[TRACE] 2017-04-25 19:38:59,888 @ qz.printer.action.WebApp:?
	Setting HTML page height to 2082.7999687194824
[DEBUG] 2017-04-25 19:39:00,004 @ qz.printer.action.WebApp:?
	Attempting image capture
[DEBUG] 2017-04-25 19:39:00,765 @ qz.printer.action.PrintHTML:?
	Parsed 1 html records
[TRACE] 2017-04-25 19:39:00,773 @ qz.printer.action.PrintPixel:?
	DPI: 609.6	CNV: 2.83464
[TRACE] 2017-04-25 19:39:00,773 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:215.90044,279.40054
[TRACE] 2017-04-25 19:39:00,773 @ qz.printer.action.PrintPixel:?
	[60900x60900 dphi, (0.0,0.0)->(215.9,279.401)mm, color]
[INFO] 2017-04-25 19:39:00,780 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-04-25 19:39:00,780 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-04-25 19:39:00,812 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-04-25 19:39:00,813 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-04-25 19:39:00,814 @ qz.printer.action.PrintImage:?
	Scaling image up by x1.0
[DEBUG] 2017-04-25 19:39:00,870 @ qz.printer.action.PrintImage:?
	Paper area: 0,0:612,792
[TRACE] 2017-04-25 19:39:00,870 @ qz.printer.action.PrintImage:?
	Image size: 612.0,184.5123751628311
```

#### Units in IN

```
[TRACE] 2017-04-25 19:40:53,890 @ qz.printer.action.WebApp:?
	Zooming in by x8.333333333333334 for increased quality
[TRACE] 2017-04-25 19:40:53,890 @ qz.printer.action.WebApp:?
	Setting HTML page width to 6800.000000000001
[TRACE] 2017-04-25 19:40:53,993 @ qz.printer.action.WebApp:?
	Setting HTML page height to 2050.0
[DEBUG] 2017-04-25 19:40:54,108 @ qz.printer.action.WebApp:?
	Attempting image capture
[DEBUG] 2017-04-25 19:40:54,867 @ qz.printer.action.PrintHTML:?
	Parsed 1 html records
[TRACE] 2017-04-25 19:40:54,874 @ qz.printer.action.PrintPixel:?
	DPI: 600.0	CNV: 72.0
[TRACE] 2017-04-25 19:40:54,874 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2017-04-25 19:40:54,875 @ qz.printer.action.PrintPixel:?
	[60000x60000 dphi, (0.0,0.0)->(215.9,279.4)mm, color]
[INFO] 2017-04-25 19:40:54,881 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-04-25 19:40:54,881 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-04-25 19:40:54,931 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-04-25 19:40:54,932 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-04-25 19:40:54,933 @ qz.printer.action.PrintImage:?
	Scaling image up by x1.0
[DEBUG] 2017-04-25 19:40:54,985 @ qz.printer.action.PrintImage:?
	Paper area: 0,0:612,792
[TRACE] 2017-04-25 19:40:54,986 @ qz.printer.action.PrintImage:?
	Image size: 612.0,184.5
```

#### Units in MM and `density: 23.622`
```
[WARN] 2017-04-25 20:28:42,784 @ qz.printer.PrintOptions:?
	Cannot read null as a double for fallbackDensity, using default
[TRACE] 2017-04-25 20:28:42,978 @ qz.printer.action.WebApp:?
	Waiting for capture..
[TRACE] 2017-04-25 20:28:42,999 @ qz.printer.action.WebApp:?
	Done: 100.0 > -1.0
[TRACE] 2017-04-25 20:28:43,000 @ qz.printer.action.WebApp:?
	New state: SUCCEEDED > READY
[TRACE] 2017-04-25 20:28:43,001 @ qz.printer.action.WebApp:?
	Done: -1.0 > 0.0
[TRACE] 2017-04-25 20:28:43,001 @ qz.printer.action.WebApp:?
	New state: READY > SCHEDULED
[TRACE] 2017-04-25 20:28:43,001 @ qz.printer.action.WebApp:?
	New state: SCHEDULED > RUNNING
[TRACE] 2017-04-25 20:28:43,008 @ qz.printer.action.WebApp:?
	Done: 0.0 > 100.0
[TRACE] 2017-04-25 20:28:43,009 @ qz.printer.action.WebApp:?
	New state: RUNNING > SUCCEEDED
[TRACE] 2017-04-25 20:28:43,011 @ qz.printer.action.WebApp:?
	Zooming in by x8.333316541512808 for increased quality
[TRACE] 2017-04-25 20:28:43,011 @ qz.printer.action.WebApp:?
	Setting HTML page width to 6799.986297874451
[TRACE] 2017-04-25 20:28:43,128 @ qz.printer.action.WebApp:?
	Setting HTML page height to 2049.995869212151
[DEBUG] 2017-04-25 20:28:43,244 @ qz.printer.action.WebApp:?
	Attempting image capture
[DEBUG] 2017-04-25 20:28:43,981 @ qz.printer.action.PrintHTML:?
	Parsed 1 html records
[TRACE] 2017-04-25 20:28:43,988 @ qz.printer.action.PrintPixel:?
	DPI: 599.9988	CNV: 2.83464
[TRACE] 2017-04-25 20:28:43,988 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:215.90044,279.40054
[TRACE] 2017-04-25 20:28:43,989 @ qz.printer.action.PrintPixel:?
	[59900x59900 dphi, (0.0,0.0)->(215.9,279.401)mm, color]
[INFO] 2017-04-25 20:28:44,001 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-04-25 20:28:44,001 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-04-25 20:28:44,145 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-04-25 20:28:44,145 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-04-25 20:28:44,147 @ qz.printer.action.PrintImage:?
	Scaling image up by x1.0
[DEBUG] 2017-04-25 20:28:44,259 @ qz.printer.action.PrintImage:?
	Paper area: 0,0:612,792
[TRACE] 2017-04-25 20:28:44,259 @ qz.printer.action.PrintImage:?
	Image size: 612.0,184.5
```

**Comment by tresf** [comment]
@klabarge thanks for testing.

The documentation for [`PrinterResolution` (`ResolutionSyntax`)](https://docs.oracle.com/javase/7/docs/api/javax/print/attribute/ResolutionSyntax.html) suggests this is a bad candidate due to rounding problems.

> "... roundoff error will in general occur if a resolution attribute's values are created in one units and retrieved in different units; for example, 600 dpi will be rounded to 236 dpcm, whereas the true value (to five figures) is 236.22 dpcm."

I'll defer to @bberenz for a proper technique.

**Comment by bberenz** [comment]
4b6fb90a7519c92ede36a57741c8f55e89abff4a should clear this up
We are just going to bypass PrinterResolution class's conversion methods so as to avoid their integer down-casting.

**Comment by klabarge** [comment]
Yes, this patch works. I tested `mm` and the DPI is correctly set to 600.

```
	DPI: 600.0	CNV: 2.83464
[TRACE] 2017-05-01 22:43:33,078 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:209.9032,297.0395
[TRACE] 2017-05-01 22:43:33,078 @ qz.printer.action.PrintPixel:?
	[color, (0.0,0.0)->(209.903,297.04)mm, 60000x60000 dphi]
[INFO] 2017-05-01 22:43:33,084 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-05-01 22:43:33,084 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-05-01 22:43:33,119 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-05-01 22:43:33,123 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-05-01 22:43:33,127 @ qz.printer.action.PrintImage:?
	Scaling image up by x1.0
[DEBUG] 2017-05-01 22:43:33,204 @ qz.printer.action.PrintImage:?
	Paper area: 0,0:595,842
[TRACE] 2017-05-01 22:43:33,204 @ qz.printer.action.PrintImage:?
	Image size: 595.0,184.4751966122202
```
